Source/WebCore/dom/ElementTextDirection.cpp

```diff
@@ -0,0 +1,325 @@
+{
+    if (equalLettersIgnoringASCIICase(value, "ltr"_s))
+        return TextDirectionality::LTR;
+    if (equalLettersIgnoringASCIICase(value, "rtl"_s))
+        return TextDirectionality::RTL;
+    if (equalLettersIgnoringASCIICase(value, "auto"_s))
+        return TextDirectionality::Auto;
+    return TextDirectionality::Undefined;
+}
+
+static TextDirectionality elementTextDirectionality(const Element& element)
+{
+    return parseTextDirectionality(element.attributeWithoutSynchronization(dirAttr));
+}
+
+bool elementHasValidTextDirectionality(const Element& element)
+{
+    return elementTextDirectionality(element) != TextDirectionality::Undefined;
+}
+
+static bool elementHasOwnValidTextDirectionality(const Element& element)
+{
+    return is<HTMLBDIElement>(element) || elementHasValidTextDirectionality(element);
+}
+
+bool elementHasOwnValidAutoTextDirectionality(const Element& element)
+{
+    auto textDirectionality = elementTextDirectionality(element);
+    return textDirectionality == TextDirectionality::Auto || (is<HTMLBDIElement>(element) && textDirectionality == TextDirectionality::Undefined);
+}
+
+static void setNodeHasDirAutoFlagRecursively(Node& firstNode, bool flag)
+{
+    firstNode.setSelfOrPrecedingNodesAffectDirAuto(flag);
+
+    for (RefPtr node = firstNode.firstChild(); node; ) {
+        auto* element = dynamicDowncast<Element>(*node);
+
+        if (element && elementHasOwnValidTextDirectionality(*element)) {
+            node = NodeTraversal::nextSkippingChildren(*node, &firstNode);
+            continue;
+        }
+
+        node->setSelfOrPrecedingNodesAffectDirAuto(flag);
+        node = NodeTraversal::next(*node, &firstNode);
+    }
+}
+
+static void setElementHasDirAutoFlag(Element& element, TextDirectionality textDirectionality)
+{
+    RefPtr parent = element.parentOrShadowHostElement();
+    element.protectedDocument()->setIsDirAttributeDirty();
+
+    switch (textDirectionality) {
+    case TextDirectionality::LTR:
+    case TextDirectionality::RTL:
+        if (element.selfOrPrecedingNodesAffectDirAuto() || (parent && parent->selfOrPrecedingNodesAffectDirAuto()))
+            setNodeHasDirAutoFlagRecursively(element, false);
+        break;
+
+    case TextDirectionality::Auto:
+        setNodeHasDirAutoFlagRecursively(element, true);
+        break;
+
+    case TextDirectionality::Undefined:
+        if (element.selfOrPrecedingNodesAffectDirAuto() && !(parent && parent->selfOrPrecedingNodesAffectDirAuto()) && !is<HTMLBDIElement>(element))
+            setNodeHasDirAutoFlagRecursively(element, false);
+        break;
+    }
+}
+
+static std::optional<TextDirection> computeTextTextDirection(const String& text)
+{
+    if (auto direction = text.defaultWritingDirection())
+        return *direction == U_LEFT_TO_RIGHT ? TextDirection::LTR : TextDirection::RTL;
+    return std::nullopt;
+}
+
+static std::optional<TextDirection> computeElementDescendantsAutoTextDirection(const Element& element)
```

Comment:
This function computes the text direction of element from text content, we should call it that:
e.g. computeTextDirectionFromText
We should place a URL to where this is spec'ed

Reply:
The function was renamed to `computeTextDirectionFromText()` and a URL to the computation algorithm was added to `computeElementTextDirection()`. Unfortunately the sections of the algorithm can't be referenced individually.

Reply:
A link to the specs was added also.